compiler/src/iree/compiler/Codegen/TransformDialectStrategies/Common/Common.cpp

```diff
@@ -193,19 +193,23 @@ mlir::iree_compiler::buildTileFuseToScfFor(ImplicitLocOpBuilder &b,
 /// If `resultingFusedOpsHandles` is a non-null pointer, the fused operation are
 /// appended in order.
 // TODO: apply forwarding pattern.
-template <typename TilingTransformOp, typename TileOrNumThreadSpec>
+template <typename TileOrNumThreadSpec>
 static iree_compiler::TileToForallAndFuseAndDistributeResult
 buildTileAndFuseAndDistributeImpl(ImplicitLocOpBuilder &b,
                                   Value isolatedParentOpH, Value rootH,
                                   ValueRange opsHToFuse,
                                   ArrayRef<OpFoldResult> tileSizesOrNumThreads,
                                   ArrayAttr threadDimMapping) {
   iree_compiler::TileToForallAndFuseAndDistributeResult result;
-  auto tileToForeachOp = b.create<TilingTransformOp>(
+  auto tileToForeachOp = b.create<TileToForallOp>(
       rootH, tileSizesOrNumThreads, TileOrNumThreadSpec(), threadDimMapping);
   result.forallH = tileToForeachOp.getForallOp();
   result.tiledOpH = tileToForeachOp.getTiledOp();
 
+  // Handle the workgroup count region.
+  b.create<IREEPopulateWorkgroupCountRegionUsingNumThreadsSliceOp>(
+      result.forallH);
```

Comment:
This needs to account for the `threadDimMapping` as well. For example if I have
```
%forall_op, %tiled_op = transform.structured.tile_to_forall_op %first   num_threads [] tile_sizes [1, 16, 32](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> () 
```
which produces something like this
```
  ^bb0(%arg0: !hal.device):
    %c16200 = arith.constant 16200 : index
    %c1 = arith.constant 1 : index
    hal.return %c16200, %c1, %c1 : index, index, index
  }
...
%6 = scf.forall (%arg0, %arg1, %arg2) in (1, 1, 16200) shared_outs(%arg3 = %5) -> (tensor<1x16x518400xf16>)
...
} {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
```
which won't map the `16200` iterator to the right block id (z).

Reply:
I don't think I fully follow. Can recommend a fix based on the current implementation

Reply:
If I specify `[#gpu.block<z>, #gpu.block<y>, #gpu.block<x>]` for the block id mapping then it works fine. The problem is that the `populate_workgroup_count_region_using_num_threads_slice` op didn't seem to look at the mapping of the loop (perhaps I am missing something somewhere else regarding usage here).

Reply:
ok I see this comment
```
  // Since the workgroup count at HAL level is in x, y, z form, process the
  // workload in reverse.
```
so is the choice of mapping `[#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]` supposed to be invalid then?

Reply:
@MaheshRavishankar This example illustrates the problem
```
executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
#pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>

hal.executable private @matmul_static_dispatch_0 {
  hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
  
    hal.executable.export public @elementwise_out_of_order_block_id ordinal(0) layout(#pipeline_layout) {
    // CHECK-LABEL: hal.executable.export public @elementwise_out_of_order_block_id
    // CHECK-DAG: %[[C3:.*]] = arith.constant 3 : index
    // CHECK-DAG: %[[C5:.*]] = arith.constant 5 : index
    // CHECK-DAG: %[[C8:.*]] = arith.constant 8 : index
    // CHECK: hal.return %[[C3]], %[[C5]], %[[C8]] : index, index, index
    ^bb0(%arg0: !hal.device):
      %x, %y, %z = flow.dispatch.workgroup_count_from_body_slice
      hal.return %x, %y, %z : index, index, index
    }   
  
    builtin.module {
      func.func @elementwise_out_of_order_block_id() {
        // CHECK-LABEL: func.func @elementwise_out_of_order_block_id
        // CHECK: = scf.forall (%[[IV:.*]]) in (3, 5, 8) shared_outs(%{{.*}}) -> (tensor<3x5x8xf32>) {
        // CHECK: } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
        %c0 = arith.constant 0 : index
        %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>>
        %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
        %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>> -> tensor<3x5x8xf32>
        %empty = tensor.empty() : tensor<3x5x8xf32>
        %3 = linalg.generic {
          indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>,
                           affine_map<(d0, d1, d2) -> (d0, d1, d2)>],
          iterator_types = ["parallel", "parallel", "parallel"]}
          ins(%2 : tensor<3x5x8xf32>) outs(%empty : tensor<3x5x8xf32>) {   
          ^bb0(%in: f32, %in_0: f32):
            %4 = math.sqrt %in : f32 
            linalg.yield %4 : f32 
          } -> tensor<3x5x8xf32>
        flow.dispatch.tensor.store %3, %1, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : tensor<3x5x8xf32> -> !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
        return
      }   
    }   
  }
}

transform.sequence failures(propagate) {
^bb1(%variant_op: !pdl.operation):
  %1 = transform.structured.match ops{["linalg.generic"]} in %variant_op : (!pdl.operation) -> !pdl.operation
  %forall_op, %tiled_op = transform.structured.tile_to_forall_op %1   num_threads [] tile_sizes [1, 1, 1](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
  transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> ()
}
```
Essentially this currently produces
```
hal.return %c8, %c5, %c3 : index, index, index
```
when I think it should produce
```
hal.return %c3, %c5, %c8 : index, index, index
```
to map the iteration order specified in the loop
```
%4 = scf.forall (%arg0, %arg1, %arg2) in (3, 5, 8) shared_outs(%arg3 = %3) -> (tensor<3x5x8xf32>) {
} {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
```

Reply:
> @MaheshRavishankar This example illustrates the problem
> 
> ```
> executable_target_embedded_elf_x86_64_ = #hal.executable.target<"llvm-cpu", "embedded-elf-x86_64", {cpu = "generic", cpu_features = "", data_layout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", native_vector_size = 16 : index, target_triple = "x86_64-unknown-unknown-eabi-elf"}>
> #pipeline_layout = #hal.pipeline.layout<push_constants = 0, sets = [<0, bindings = [<0, storage_buffer, ReadOnly>, <1, storage_buffer, ReadOnly>, <2, storage_buffer>]>]>
> 
> hal.executable private @matmul_static_dispatch_0 {
>   hal.executable.variant public @embedded_elf_x86_64, target = #executable_target_embedded_elf_x86_64_ {
>   
>     hal.executable.export public @elementwise_out_of_order_block_id ordinal(0) layout(#pipeline_layout) {
>     // CHECK-LABEL: hal.executable.export public @elementwise_out_of_order_block_id
>     // CHECK-DAG: %[[C3:.*]] = arith.constant 3 : index
>     // CHECK-DAG: %[[C5:.*]] = arith.constant 5 : index
>     // CHECK-DAG: %[[C8:.*]] = arith.constant 8 : index
>     // CHECK: hal.return %[[C3]], %[[C5]], %[[C8]] : index, index, index
>     ^bb0(%arg0: !hal.device):
>       %x, %y, %z = flow.dispatch.workgroup_count_from_body_slice
>       hal.return %x, %y, %z : index, index, index
>     }   
>   
>     builtin.module {
>       func.func @elementwise_out_of_order_block_id() {
>         // CHECK-LABEL: func.func @elementwise_out_of_order_block_id
>         // CHECK: = scf.forall (%[[IV:.*]]) in (3, 5, 8) shared_outs(%{{.*}}) -> (tensor<3x5x8xf32>) {
>         // CHECK: } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
>         %c0 = arith.constant 0 : index
>         %0 = hal.interface.binding.subspan set(0) binding(0) type(storage_buffer) alignment(64) offset(%c0) flags(ReadOnly) : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>>
>         %1 = hal.interface.binding.subspan set(0) binding(1) type(storage_buffer) alignment(64) offset(%c0) : !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
>         %2 = flow.dispatch.tensor.load %0, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : !flow.dispatch.tensor<readonly:tensor<3x5x8xf32>> -> tensor<3x5x8xf32>
>         %empty = tensor.empty() : tensor<3x5x8xf32>
>         %3 = linalg.generic {
>           indexing_maps = [affine_map<(d0, d1, d2) -> (d0, d1, d2)>,
>                            affine_map<(d0, d1, d2) -> (d0, d1, d2)>],
>           iterator_types = ["parallel", "parallel", "parallel"]}
>           ins(%2 : tensor<3x5x8xf32>) outs(%empty : tensor<3x5x8xf32>) {   
>           ^bb0(%in: f32, %in_0: f32):
>             %4 = math.sqrt %in : f32 
>             linalg.yield %4 : f32 
>           } -> tensor<3x5x8xf32>
>         flow.dispatch.tensor.store %3, %1, offsets = [0, 0, 0], sizes = [3, 5, 8], strides = [1, 1, 1] : tensor<3x5x8xf32> -> !flow.dispatch.tensor<writeonly:tensor<3x5x8xf32>>
>         return
>       }   
>     }   
>   }
> }
> 
> transform.sequence failures(propagate) {
> ^bb1(%variant_op: !pdl.operation):
>   %1 = transform.structured.match ops{["linalg.generic"]} in %variant_op : (!pdl.operation) -> !pdl.operation
>   %forall_op, %tiled_op = transform.structured.tile_to_forall_op %1   num_threads [] tile_sizes [1, 1, 1](mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>])
>   transform.iree.populate_workgroup_count_region_using_num_threads_slice %forall_op : (!pdl.operation) -> ()
> }
> ```
> 
> Essentially this currently produces
> 
> ```
> hal.return %c8, %c5, %c3 : index, index, index
> ```
> 
> when I think it should produce
> 
> ```
> hal.return %c3, %c5, %c8 : index, index, index
> ```
> 
> to map the iteration order specified in the loop
> 
> ```
> %4 = scf.forall (%arg0, %arg1, %arg2) in (3, 5, 8) shared_outs(%arg3 = %3) -> (tensor<3x5x8xf32>) {
> } {mapping = [#gpu.block<x>, #gpu.block<y>, #gpu.block<z>]}
> ```

@qedawkins Thanks! There was no lit test for this in IREE. We have a definite testing hole of these paths in IREE. I fixed the issue and added a test. Please take a look.